File: python/20_concurrency.py
# 20 concurrency (threads vs async vs processes)

import threading
import time

def work(name):
    print(f"{name} start")
    time.sleep(0.3)
    print(f"{name} done")

threads = [threading.Thread(target=work, args=(f"t{i}",)) for i in range(3)]
for t in threads:
    t.start()
for t in threads:
    t.join()

print("all threads done")
print("note: use asyncio for I/O-bound, multiprocessing for CPU-bound")

File: go/20_concurrency.go
package main

import (
	"fmt"
	"sync"
	"time"
)

func work(name string, wg *sync.WaitGroup) {
	defer wg.Done()
	fmt.Println(name, "start")
	time.Sleep(300 * time.Millisecond)
	fmt.Println(name, "done")
}

func main() {
	var wg sync.WaitGroup
	for i := 0; i < 3; i++ {
		wg.Add(1)
		go work(fmt.Sprintf("g%d", i), &wg)
	}
	wg.Wait()
	fmt.Println("all goroutines done")
	fmt.Println("Go: goroutines + channels are the default concurrency model")
}
